core/src/main/java/io/questdb/network/IODispatcherLinux.java

```diff
@@ -45,66 +53,221 @@ public void close() {
+            } else {
+                final long eventId = pendingEvents.get(eventRow, EVM_ID);
+                if (epoll.control(suspendEvent.getFd(), eventId, EpollAccessor.EPOLL_CTL_DEL, 0) < 0) {
+                    LOG.error().$("epoll_ctl remove suspend event failure [eventId=").$(eventId)
+                            .$(", err=").$(nf.errno()).I$();
+                }
+                pendingEvents.deleteRow(eventRow);
+            }
+        }
+        doDisconnect(context, reason);
+    }
+
     private void enqueuePending(int watermark) {
         for (int i = watermark, sz = pending.size(), offset = 0; i < sz; i++, offset += EpollAccessor.SIZEOF_EVENT) {
-            epoll.setOffset(offset);
+            final long id = pending.get(i, OPM_ID);
             if (
                     epoll.control(
-                            (int) pending.get(i, M_FD),
-                            pending.get(i, M_ID),
+                            (int) pending.get(i, OPM_FD),
+                            id,
                             EpollAccessor.EPOLL_CTL_ADD,
                             initialBias == IODispatcherConfiguration.BIAS_READ ? EpollAccessor.EPOLLIN : EpollAccessor.EPOLLOUT
                     ) < 0
             ) {
-                LOG.debug().$("epoll_ctl failure ").$(nf.errno()).$();
+                LOG.error().$("epoll_ctl failure [id=").$(id)
+                        .$(", err=").$(nf.errno()).I$();
             }
         }
     }
 
-    private void processIdleConnections(long deadline) {
+    private boolean handleSocketOperation(long id) {
+        // find row in pending for two reasons:
+        // 1. find payload
+        // 2. remove row from pending, remaining rows will be timed out
+        final int row = pending.binarySearch(id, OPM_ID);
+        if (row < 0) {
+            LOG.error().$("internal error: epoll returned unexpected id [id=").$(id).I$();
+            return false;
+        }
+
+        final C context = pending.get(row);
+        final SuspendEvent suspendEvent = context.getSuspendEvent();
+        if (suspendEvent != null) {
+            // the operation is suspended, check if we have a client disconnect
+            if (testConnection(context.getFd())) {
+                doDisconnect(context, id, DISCONNECT_SRC_PEER_DISCONNECT);
+                pending.deleteRow(row);
+                return true;
+            }
+        } else {
+            publishOperation(
+                    (epoll.getEvent() & EpollAccessor.EPOLLIN) > 0 ? IOOperation.READ : IOOperation.WRITE,
+                    context
+            );
+            pending.deleteRow(row);
+            return true;
+        }
+        return false;
+    }
+
+    private void handleSuspendEvent(long id) {
+        final int eventsRow = pendingEvents.binarySearch(id, EVM_ID);
+        if (eventsRow < 0) {
+            LOG.error().$("internal error: epoll returned unexpected event id [eventId=").$(id).I$();
```

Comment:
should be critical

Reply:
Changed in bbec37baf832db9a846d04e59ab124f8bb84eadc